File: guet/steps/step.py
from typing import List


class Step:

    def __init__(self):
        self._next = None

    def do_play(self, args: List[str]):
        raise NotImplementedError()

    def play(self, args: List[str]):
        self.do_play(args)
        if self._next is not None:
            self._next.play(args)

    def next(self, next_step: "Step") -> "Step":
        if self._next is None:
            self._next = next_step
        else:
            self._next.next(next_step)
        return self

File: src/steps/step.ts
import { Args } from '../args';

export abstract class Step {
    private _next?: Step;
    constructor() {}

    public next(next: Step): Step {
        this._next = next;
        return this;
    }

    public abstract doPlay(args: Args): void;

    public play(args: Args) {
        this.doPlay(args);
        if (this._next) {
            this._next.play(args);
        }
    }
}
